### PR TITLE
Adding code and documentation standards

### PR DESCRIPTION
**Pull request by arschles**
This is a start to our code and documentation standards. By the end of 0.0.3, our goal is not to make this a comprehensive or exhaustive doc. Please review accordingly.

Fixes https://github.com/kubernetes-incubator/service-catalog/issues/636

### CODE REVIEW

File: docs/code-standards.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
+# Code & Documentation Standards
+
+This document details our goals for code quality, coverage and documentation standards. At the time
+of this writing, these are standards that all of our code may not meet. We should always try to 
+improve our code to meet them.
+
+## Code Quality And Documentation Standards
```
Comment (MHBauer) [inline docs/code-standards.md]: Mention `make verify`?

File: docs/code-standards.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+We aim to write clear code that is easy to read and comprehend. In particular, that goal means the
+following:
+
+1. All Go code must pass the Go linter checks
+    - We have these checks integrated into our CI system
+2. All Go code should pass the `go vet`
+    - These checks are not integrated into our CI system
```
Comment (MHBauer) [inline docs/code-standards.md]: I thought `vet` was run under `make verify`?

File: docs/code-standards.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    - Anyone with the Go toolchain installed on their system can run the vet checks by executing the
+    following command on their machine:
+
+    ```go
+    go vet ./pkg/... ./cmd/... ./test/...
+    ```
+3. All Go code must be formatted with [gofmt](https://golang.org/cmd/gofmt/)
+    - Anyone can run the formatter by runinng `make format` on their machine
+4. Any exported symbols (`type`s, `interface`s, `func`s, `struct`s, etc...) must have Godoc
```
Comment (MHBauer) [inline docs/code-standards.md]: This is pretty much a result of the linter. Do we need to mention it?

File: docs/code-standards.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+compatible comments associated with them
+5. Unexported symbols must be commented sufficiently to provide direction & context to a developer
```
Comment (MHBauer) [inline docs/code-standards.md]: Agree.

File: docs/code-standards.md [hunk 5]
```diff
@@ -0,0 +1,35 @@
+# Code & Documentation Standards
+
+This document details our goals for code quality, coverage and documentation standards. At the time
+of this writing, these are standards that all of our code may not meet. We should always try to 
+improve our code to meet them.
```
Comment (vaikas-google) [inline docs/code-standards.md]: 'our code' is not exactly right since there's also the documentation to consider.
maybe: our code and documentation

Comment (vaikas-google) [inline docs/code-standards.md]: Or is this only meant to talk about documenting the code?

File: docs/code-standards.md [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+## Code Quality And Documentation Standards
+
+We aim to write clear code that is easy to read and comprehend. In particular, that goal means the
```
Comment (pmorie) [inline docs/code-standards.md]: Probably want to link to https://github.com/kubernetes/community/blob/master/contributors/devel/coding-conventions.md here

File: docs/code-standards.md (hunk 2 above)
Comment (krancour) [inline docs/code-standards.md]: It is.

File: docs/code-standards.md [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+following:
+
+1. All Go code must pass the Go linter checks
+    - We have these checks integrated into our CI system
+2. All Go code should pass the `go vet`
```
Comment (krancour) [inline docs/code-standards.md]: "the `go vet`" reads oddly. Suggest removing "the".

File: docs/code-standards.md (hunk 2 above)
Comment (krancour) [inline docs/code-standards.md]: While I wouldn't suggest removing (1) and (2) above, I agree with what @MHBauer mentions [here](https://github.com/kubernetes-incubator/service-catalog/pull/696/files#r111459838). _All_ such checks are bundled together in `make verify`, so if for no reason other than convenience, we should mention that.

File: docs/code-standards.md (hunk 3 above)
Comment (krancour) [inline docs/code-standards.md]: Does it matter whether it comes from `gofmt`, `go vet`, or `golint`? This is widely held as a best practice and I know @arschles and myself are both sticklers about this. Even if no tool complained about this, I'd _still_ want to mention it.

Comment (arschles) [inline docs/code-standards.md]: Yes, we need to mention it

**Comment by arschles** [comment]
@MHBauer @vaikas-google PTAL

Also, @MHBauer I have linked to this from the dev guide

**Comment by krancour** [comment]
@arschles, it's not blocking my LGTM, which I will add now, but if we could mention `make verify`, that is the painless way of asserting we pass all the checks you've mentioned.

File: docs/code-standards.md [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    - Anyone can run the formatter by runinng `make format` on their machine
+4. Any exported symbols (`type`s, `interface`s, `func`s, `struct`s, etc...) must have Godoc
+compatible comments associated with them
+5. Unexported symbols must be commented sufficiently to provide direction & context to a developer 
+who didn't write the code
+6. Inline code should be commented sufficiently to explain what complex code is doing. It's up to
+the developer and reviewers how much and what kind of documentation is necessary
+7. Unit, integration or end-to-end tests should be written for all business logic
+    - Reviewers should take care to ensure the right, and enough testing is written for a PR
+    - We do not yet check code coverage in our CI system
```
Comment (simonleung8) [inline docs/code-standards.md]: Is this relevant? The PR should have all logic tested and the reviewer should make sure of that.

Comment (arschles) [inline docs/code-standards.md]: It's still true that we don't check code coverage in CI

File: docs/code-standards.md [hunk 9]
```diff
@@ -0,0 +1,39 @@
+# Code & Documentation Standards
+
+This document details our goals for code quality, coverage and documentation standards. At the time
+of this writing, these are standards that all of our code may not meet. We should always try to 
+improve our codebase and documentation to meet them.
+
+While we do not currently aim to adhere completely to the
```
Comment (pmorie) [inline docs/code-standards.md]: Why not?

Comment (arschles) [inline docs/code-standards.md]: Because that's large document and we haven't discussed in full yet. This document may change once we do.

Comment (pmorie) [inline docs/code-standards.md]: OK

**Pull request merged**